libs/aio/odc/aio/__init__.py

```diff
@@ -292,6 +292,82 @@ def default_guide(url, depth, base):
     return step
 
 
+def s3_uri_to_stream(uri: str, skip_check: bool):
+    """Build generator from supplied S3 URI pattern
+    Arguments:
+        uri {str} -- S3 URI to create generator for
+        skip_check {bool} -- Validity check flag
+    Raises:
+        e: ValueError if parsing URI fails
+    Returns:
+        [type] -- [description]
+    Yields:
+        [type] -- [description]
+    """
+    def do_file_query(qq, pred):
+        for d in s3.dir_dir(qq.base, qq.depth):
+            _, _files = s3.list_dir(d).result()
+            for f in _files:
+                if pred(f):
+                    yield f
+
+    def do_file_query2(qq):
+        fname = qq.file
+
+        stream = s3.dir_dir(qq.base, qq.depth)
+
+        if skip_check:
+            yield from (SimpleNamespace(url=d+fname) for d in stream)
+            return
+
+        stream = (s3.head_object(d+fname) for d in stream)
+
+        for (f, _), _ in future_results(stream, 32):
+            if f is not None:
+                yield f
+
+    def do_dir_query(qq):
+        return (SimpleNamespace(url=url) for url in s3.dir_dir(qq.base, qq.depth))
+
+    try:
+        qq = parse_query(uri)
+    except ValueError as e:
+        # TODO: Log this error
+        raise e
+
+    s3 = S3Fetcher()
+
+    glob_or_file = qq.glob or qq.file
+
+    if qq.depth is None and glob_or_file is None:
+        stream = s3.find(qq.base)
+    elif qq.depth is None or qq.depth < 0:
+        if qq.glob:
+            stream = s3.find(qq.base, glob=qq.glob)
+        elif qq.file:
+            postfix = '/'+qq.file
+            stream = s3.find(qq.base, pred=lambda o: o.url.endswith(postfix))
+    else:
+        # fixed depth query
+        if qq.glob is not None:
+            pred = norm_predicate(glob=qq.glob)
+            stream = do_file_query(qq, pred)
+        elif qq.file is not None:
+            stream = do_file_query2(qq)
+        else:
+            stream = do_dir_query(qq)
+
+    yaml_list = []
+
+    try:
+        for i, o in enumerate(stream):
+            yaml_list.append(o.url)
```

Comment:
I think you don't want to return list here, instead of adding things to List just `yield o.url`
But do consider error handling as well

Reply:
Also doesn't looks like you use `i` so probably don't need `enumerate`

Reply:
The library does not use a logger infrastructure. Do you have a design in mind or should we just do logging.warning at the top level logger.

Reply:
For logging we just use python logging elsewhere. So I guess same here. But logging is separate from error handling, in this case you are just returning a lazy sequence of URIs from a single glob pattern, so there isn't much to do error handling wise. I mean underlying layers should be doing it properly though.

Next step where you actually fetch data will need it though.